anaconda_packaging_utils/api/jira_api.py

```diff
@@ -0,0 +1,76 @@
+"""
+File:           jira_api.py
+Description:    Wrapper library that provides tools for using the Python JIRA API. The primary purpose of this wrapper
+                is to simplify and standardize the authentication process.
+
+                JIRA API links:
+                  - Source: https://github.com/pycontribs/jira
+                  - Docs: https://jira.readthedocs.io/index.html#
+"""
+import logging
+from typing import Final
+
+from jira.client import JIRA
+
+from anaconda_packaging_utils.storage.config_data import ConfigData
+
+# Logging object for this module
+log = logging.getLogger(__name__)
+
+# Where our JIRA boards are hosted
+_JIRA_HOST_URL: Final[str] = "https://anaconda.atlassian.net/"
+
+
+class ApiException(Exception):
+    """
+    Generic exception indicating an unrecoverable failure of this API.
+
+    This exception is meant to condense many possible failures into one generic error. The thinking is, if the calling
+    code runs into any API failure, there isn't much that can be done. So it is easier for the caller to handle one
+    exception than many exception types.
+    """
+
+    def __init__(self, message: str):
+        """
+        Constructs an API exception
+        :param message: String description of the issue encountered.
+        """
+        super().__init__(message if len(message) else "An unknown API issue was encountered.")
+
+
+class JiraApi:
+    """
+    Singleton wrapper to the Python Jira project. This "ensures" that we only construct and authenticate the underlying
```

Comment:
I hate singletons so much as they holding internal object state with itself, have you consider using factory pattern instead ?
They are perceived often as anti pattern, see https://medium.com/aia-sg-techblog/why-singleton-pattern-is-considered-as-anti-design-pattern-c81dd8b7e757

Reply:
I guess I was trying to minimize the number of auth calls and object constructions. I think the only state that should remain the same in these APIs is the authorization state, which I think we want for our tools.

Both of these should be making stateless HTTP REST calls under the hood. So I think the Singleton pattern makes a lot of sense here. But let me know if you have specific concerns.

Reply:
The problem with singleton can start being apparent when you do tests, you returning JIRA object which would hold its state across different test cases, if you call any stateful method on JIRA object the result could be observer in follow up tests, which could break the 1st rule that the tests should be run independent from each other. The solution for ( if you still want to use singleton ) is to have a wrapper singleton in your test code which would be able to reset the underlying instance each time new test is run